### PR TITLE
[test] extend tests for immutability of maps 

### LINKED ISSUE #3724: [BUG] Side effects when calling map functions
**Describe the bug**

In some cases it is possible that the value of a variable bound to a map is mutated. 

**Expected behavior**

In XQuery, variables are immutable. Once a variable is bound, its value cannot change within the scope of its FLWOR expression and binding sequence. See https://www.w3.org/TR/xquery-31/#id-binding-rules.

In the example below, the variable `$three-entry-map` is bound to a map containing three entries:

```xquery
map {
    "a": 1,
    "b": 2,
    "c": 3
}
```

The query returns this variable—this 3-entry map—and it does so in BaseX and Saxon. 

However, in eXist, the presence of a map-altering function like `map:remove()`, `map:put()`, or `map:merge()` in a subsequent `let` clause somehow mutates the `$three-entry-map` variable. As a result, in eXist, the example queries all returns the following map, with only 2 entries instead of the expected 3:

```xquery
map {
    "a": 1,
    "c": 3
}
``` 

This violates immutability. 

**To Reproduce**

The problem has been reduced to the following:

```xquery
xquery version "3.1";

let $three-entry-map := 
    map:merge((
        map {
            "a" : 1,
            "b" : 2,
            "c" : 3
        }
    ))
let $some-other-map := map:remove($three-entry-map, "b")
return 
    $three-entry-map
```

In BaseX and Saxon, this returns the expected 3-entry map. In eXist, it returns a 2-entry map.

Similarly, the presence of `map:put()` will trigger the problem in eXist:

```xquery
xquery version "3.1";

let $three-entry-map := 
    map:merge((
        map {
            "a" : 1,
            "b" : 2,
            "c" : 3
        }
    ))
let $some-other-map := map:put($three-entry-map, "b", "foo")
return 
    $three-entry-map
```

Lastly, `map:merge()` will trigger it too:

```xquery
xquery version "3.1";

let $three-entry-map := 
    map:merge((
        map {
            "a" : 1,
            "b" : 2,
            "c" : 3
        }
    ))
let $some-other-map := 
    map:merge((
        $three-entry-map, 
        map { "b": "foo" }
    ))
return 
    $three-entry-map
```

In all 3 cases, in BaseX and Saxon, this returns the expected 3-entry map. In eXist, it returns a 2-entry map.

Even this case without a `let` clause reveals the problem - returning an empty map instead of a 1-entry map:

```xquery
xquery version "3.1";

map:merge( map { "a" : 1 } ) ! (., map:put(., "a", 2))[1]
```

 - OS: linux, windows, macOS
 - eXist-db version: 5.3.0-SNAPSHOT 303ca0db511ee93bac1b50ccede357fde9658e7b
 - Java Version 1.8.0_271

### PR DESCRIPTION
**Pull request by line-o**
### Description:

This PR now only adds tests that extend the previous test for the immutability of maps

~~fork maps before mutating them~~

### Reference:

refs #3724

Map implementation (bifurcan) documentation of the `forked()` method
https://lacuna.io/docs/bifurcan/io/lacuna/bifurcan/IMap.html#forked--

### Type of tests:

Extended XQsuite tests in exist-core/src/test/xquery/maps/maps.xql

### CODE REVIEW

**Comment by line-o** [comment]
So, in `SingleKeyMapType#put` a different method is used to create the copy of the map.
`return new MapType(context, map.forked(), keyType);`
I have not enough knowledge yet which one is the better in the instances changed here.

`.forked()` or `.clone()` - this will likely have effects on memory usage and maybe performance.

**Comment by line-o** [comment]
I do not understand the underlying implementation of "forked", though:
```java
    default IMap<K, V> forked() {
        return this;
    }
```

**Comment by line-o** [comment]
using `.forked()` also works, so we can freely choose.

**Comment by line-o** [comment]
Ready to be merged from my point of view. Tests fail until the fix is applied.

**Review by dizzzz — APPROVED** [review]

**Review by joewiz — APPROVED** [review]
I can confirm that when building eXist from this PR's branch, the tests that @PieterLamers and I developed in https://github.com/eXist-db/exist/issues/3724 all return the expected results.

File: exist-core/src/main/java/org/exist/xquery/functions/map/MapType.java
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/xquery/functions/map/MapType.java:159]: ah ok. Being not at home, I could not see the functionality behind `linear()`

**Comment by line-o** [comment]
@joewiz great!

Do you think the test cases I added are covering every combination we need to check?
https://github.com/eXist-db/exist/pull/3725/files#diff-6cbf91fe2ed15b2f66e0c96ba43b21e94f3be9075de054b215a1cb0b5a88cfd7

**Comment by joewiz** [comment]
@line-o Yes, I think you've got the ones in the issue covered with mt:immutable-put-after-merge, mt:immutable-remove-after-merge, and mt:immutable-merge-after-merge, and the other variations were a great idea too.

File: exist-core/src/main/java/org/exist/xquery/functions/map/MapType.java
Comment (adamretter) [reply exist-core/src/main/java/org/exist/xquery/functions/map/MapType.java:159]: I don't understand why forked is now used here, as `forked` is called below.

Comment (line-o) [reply exist-core/src/main/java/org/exist/xquery/functions/map/MapType.java:159]: That is too late, as the tests clearly prove. But maybe the second call to forked can be removed or changed to `.linear()`?

Comment (line-o) [reply exist-core/src/main/java/org/exist/xquery/functions/map/MapType.java:159]: The second call will likely not do much the documentation states:
"For already forked maps the call to .forked() will just return itself".

**Comment by adamretter** [comment]
@line-o I just built the HEAD of Bifurcan and checked against that too. I wanted to do so, as it has some fixes that are not released yet. Unfortunately there was no change, still 6 failing tests. I will dig a little deeper...

**Comment by adamretter** [comment]
@line-o I just expanded on some of your tests, and it caused me to find another bug in maps. The XQuery F+O specs says that default behaviour for `map:merge` when there are duplicates keys is to `use-first`, however eXist-db does `use-last` and some of your tests were relying on that - I will fix that too...

**Comment by dizzzz** [comment]
Is there any new info on this?

**Comment by dizzzz** [comment]
what is the state of the PR?  @line-o  @adamretter 
I marked it as draft to prevent accidental pulling in

**Comment by line-o** [comment]
I am waiting for this to either a) be merged as it is or for a review that tells what needs to change (I expect this to be "remove the surplus calls to forked()")
What I will definitely do is fix the minor issue in one the tests.

**Comment by line-o** [comment]
I can confirm that #3724 is fixed by the new release of bifurcan.

**Comment by joewiz** [comment]
@line-o That's great news! Thank you for all of your efforts on the maps issue, Juri! When you have the chance, could you please approve #3886?

**Comment by line-o** [comment]
Oh wait. I was under the impression that this PR was already applied!
This needs to be retested.

**Comment by line-o** [comment]
I pulled the PR with bifurcan pre6 built exist and re-run the tests. It appears to not been fixed with that.

**Comment by adamretter** [comment]
@line-o yes I know. I did the same over the weekend. I'm still working on it...

**Comment by line-o** [comment]
I would appreciate you let me finish this here. Or at least argue why this is not the way to go instead.

**Comment by line-o** [comment]
Rebase and with improved test cases.
I still see odd behaviour and fear that there is a) still issues in the map implementation or b) issues in how this library is used.

**Comment by adamretter** [comment]
This is no longer neeed. The bug was indeed in Bifurcan and was fixed as https://github.com/lacuna/bifurcan/issues/35#issuecomment-841802782 and upgraded in eXist-db as https://github.com/eXist-db/exist/pull/3886

The tests from this PR have been incorporated and expanded upon in https://github.com/eXist-db/exist/pull/3738

@line-o are you happy to close this one now?

**Comment by line-o** [comment]
@adamretter I actually re-incorporated your tests here and added a final improvement to them. What about we make this PR the one adding the tests and #3738 can then be solely about `map:merge#2`.
That is also nicer later when we compile the changelog.

**Comment by line-o** [comment]
I will take responsibility in changing the PR description and title as well as the commit history assuming you do so for #3738 as well.

**Comment by adamretter** [comment]
@line-o I don't think that makes sense, you would also need to remove all your Java changes in this PR.
As they were an early attempt to fix the Maps problem that was reported in eXist-db's code, but the real problem wasn't in eXist-db's code... it was in Bifurcan which I worked on diagnose and get it resolved at its source. Wouldn't this PR then solely end up just being tests most of which were created in my other PR?

If you want my PR split into two, with and without the change to merge-order to match the spec, I have previously already offered to do that...

**Comment by line-o** [comment]
Of course I will remove the no longer needed commits.

**Comment by line-o** [comment]
I would also rather squash all the back and forth in the tests into a single commit.

**Pull request closed without merging**
